### ISSUE TITLE
shiftFootprints on  GAlignments reads doesn't work correctly for gapped reads

### ISSUE THREAD

**Issue by Miswi** [Roleren/ORFik#97]
Hey, as in the title: I found that the function doesn't shift reads correctly, it is restricted to the leftmost part of a read.

[footprints.txt](https://github.com/Roleren/ORFik/files/4943844/footprints.txt)

library(ORFik)
library(data.table)

dataframeAsGAlignments <- function(x) {
  GAlignments(seqnames = as.factor(x$seqnames), cigar = x$cigar,pos = x$start, strand = factor(x$strand, levels = c("+", "-", "*")))
} # why GAlignments(data.frame(gal)) doesnt return initial object is not understandable, but it's a fixer. 

footprints <- read.table("footprints.txt")
footprints <- dataframeAsGAlignments(footprints)
shift12 <- shiftFootprints(footprints, data.table(fraction = 29, offsets_start = -12))
shift5 <- shiftFootprints(footprints, data.table(fraction = 29, offsets_start = -5))
identical(shift12[2], shift5[2]) # returns TRUE

**Comment by Miswi** [comment]
I needed splice-aware shifting function compatible with ORFik very badly, so I just implemented one (I didn't account for softclipping):

applyShifts <- function(footprints, shifts) {
  selected_shifts <- shifts$offsets_start
  selected_lengths <- shifts$fraction
  footprints <- footprints[qwidth(footprints) %in% selected_lengths] 
  lengthsAll <- qwidth(footprints)
  shiftsAll <- -selected_shifts[match(lengthsAll, selected_lengths)] + 1
  shiftsAll[which(footprints@strand == "-")] <- -shiftsAll[which(footprints@strand == "-")] 
  footprints <- qnarrow(footprints, shiftsAll, shiftsAll)
  footprints <- GRanges(footprints)
  footprints$size <- lengthsAll
  return(footprints)
  }

It shifts reads correctly, but brings another issue; ORFik is very powerful and as stated in the documentation:
"The focus of ORFik for development is to be a Swiss army knife for transcriptomics"
Which it has a potential to become. Using ideas from ORFik and development in accordance to it is, however, somewhat impeded by strict use of standard functions rather than class-specific methods of defined generic functions. 
What I mean is that ideally there would be "shiftFootprints" generic function defined in the shiftFootprints.R script. In the case of bioconductor core packages its syntax is universally:
setGeneric("newfunction", function(x) standardGeneric("newfunction"))
But it leaves little room for applying the function for classes on which no method of such name is defined (i.e. if you want exactly the same function deployed for more than one class, you have to set a method for each class)
It can be dealt with by defining general function and then referring to it in multiple short methods or defining default function, in this case it would be:
setGeneric("shiftFootprints", function(footprints, shifts) {your function})
The benefit of doing it this way is that any further developer can easily modify/reimplement your function for novel, package-specific, classes; in my case, I would simply have to:
setMethod("shiftFootprints", signature("GAlignments", "data.table"),
function(x,y) applyShifts(x,y))
Thus avoiding redundant naming (it can be really hard to get grasp of a new package when all function names are new, while these could be simply methods named as usual but defined on a new class)
There is more on that here: 
http://adv-r.had.co.nz/S4.html

Sorry for giving advices/suggestions - I realize rebuilding the package so it follows structured generic functions definitions and setting methods for each class is a lot of hassle at this point, but it makes it much easier to develop based on the package down the line, what I think will be very important in this case. ORFik extends capabilities of bioconductor so much that it honestly should become one its core package. 

Best wishes, 
Michał

**Comment by Miswi** [comment]
Another issue I noticed is that shiftFootprints doesnt work on GRanges objects, only on GAlignments (so far ungapped), yet the only way to dereplicate reads (collapse) is convertToOneBasedRanges

**Comment by Roleren** [comment]
Working on this now, will see if it is not correct.

About ORFik as part of bioconductor core, I am talking to Martin Morgan about this, but not sure if I should subset part of the package out then, but that is up to him what he thinks is best.

**Comment by Roleren** [comment]
You are correct, ORFik is supposed to handle spliced / gapped reads and this was purely a test I made that should not have been updated into branch, since it used to use qnarrow which is the correct function. 

I will update it now, and add a sort argument, so you can choose to sort output or not, for easier validation of bugs. 

Should have a new update by 30 minutes.

**Comment by Roleren** [comment]
Check out latest branch on github, should now work as expected. Let me know if there is anything else relating to this issue.

**Comment by Miswi** [comment]
I’ll be back from vacation on Thursday, will check it out then.
